fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,480 @@
+        self.fc.weight = proj_weight
+        assert self.fc.weight.dtype in [torch.float16, torch.float32], self.fc.weight.dtype
+        self.fp16 = self.fc.weight.dtype == torch.float16
+        self.log = log
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        """ Forward function that computes softmax output with the input and target."""
+        assert kwargs == {}
+        input, target = input
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        if self.fp16:
+            assert input.dtype == torch.float16
+        x = self.fc(input)
+        # Note that we do softmax in FP32, which is important for numerical stability.
+        if self.log:
+            x = F.log_softmax(x, dim=-1, dtype=torch.float32)
+        else:
+            x = F.softmax(x, dim=-1, dtype=torch.float32)
+        assert x.dtype == torch.float32
+        return x
+
+
+class BaselineSoftmaxNllLoss(BaselineSoftmax):
+    """ Baseline that does an output projection, a softmax & a NLL loss (cross-entropy).
+
+        See BaselineSoftmax above. Constructor is the same. Only difference is in the
+        forward function.
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        """Forward that directly compute the loss."""
+        assert kwargs == {}
+        input, target = input
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        x = super().forward(input, target)
+        x = F.nll_loss(x, target, reduction="sum")
+        return x
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
+
+    @staticmethod
+    def get_max(i: torch.Tensor, w: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _m = torch.matmul(i, w.T)
+        if fp:
+            _m = _m.float()
+        _m = _m.max(dim=1)[0]
+        return _m
+
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any,
+        i: torch.Tensor,
+        w: torch.Tensor,
+        kernel_obj: "MemoryEfficientVocabOutput",
+        w_idx: int,
+        w_split_size: int,
+        split_dim: int,
+    ) -> torch.Tensor:
+        """Forward function that computes the max, without saving activations."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG max fwd")
+        ctx.save_for_backward(i, w)
+        ctx.kernel_obj = kernel_obj
+        ctx.w_idx = w_idx
+        ctx.w_split_size = w_split_size
+        ctx.args = {}
+        assert split_dim == 0
+        with torch.no_grad():
```

Comment:
this is to save some memory/speed, because autograd is wrong there ? curious as of why this is needed basically

Reply:
This is for saving memory with checkpointing. I will have an extensive doc that describes the algorithm in more details. Will add more comment here too.